ci-operator/step-registry/idp/freeipa/idp-freeipa-commands.sh

```diff
@@ -0,0 +1,403 @@
+                "preferredUsername": ["uid"]
+            },
+            "ca": {
+                "name": "ipa-ca"
+            },
+            "insecure": false,
+            "url": "ldaps://freeipa.freeipa.freeipa.svc.cluster.local/cn=users,cn=accounts,dc=ipa,dc=example,dc=com?uid"
+        },
+        "mappingMethod": "claim",
+        "name": "IPA",
+        "type": "LDAP"
+    }'
+
+    # Since no IDP is configured, create a new array with the IPA IDP
+    combined_idp_config="[$new_ipa_idp]"
+
+    # Construct the JSON patch using a here-document
+    json_patch=$(cat <<EOF
+{
+  "spec": {
+    "identityProviders": $combined_idp_config
+  }
+}
+EOF
+)
+
+    # Apply the patch
+    if ! oc patch oauth cluster --type=merge -p "$json_patch"; then
+        echo "Failed to apply the patch to configure FreeIPA IDP"
+        exit 1
+    fi
+
+    # wait for oauth-openshift to rollout
+    wait_auth=true
+    expected_replicas=$(oc get deployment oauth-openshift -n openshift-authentication -o jsonpath='{.spec.replicas}')
+    while $wait_auth;
+    do
+        available_replicas=$(oc get deployment oauth-openshift -n openshift-authentication -o jsonpath='{.status.availableReplicas}')
+        new_gen=$(oc get deployment oauth-openshift -n openshift-authentication -o jsonpath='{.metadata.generation}')
+        if [[ $expected_replicas == "$available_replicas" && $((new_gen)) -gt $((gen)) ]]; then
+            wait_auth=false
+        else
+            sleep 10
+        fi
+    done
+
+    # store users in a shared file
+    if [ -f "${SHARED_DIR}/runtime_env" ] ; then
+        source "${SHARED_DIR}/runtime_env"
+    fi
+    runtime_env=${SHARED_DIR}/runtime_env
+    users=${users::-1}
+
+
+    cat <<EOF >>"${runtime_env}"
+export USERS=${users}
+EOF
+}
+
+# Error handling function
+function handle_error() {
+    echo "Error occurred in $1"
+    exit 1
+}
+
+# Check cluster type and skip steps if necessary
+if [ -f "${SHARED_DIR}/cluster-type" ] ; then
+    CLUSTER_TYPE=$(cat "${SHARED_DIR}/cluster-type")
+    if [[ "$CLUSTER_TYPE" == "osd" ]] || [[ "$CLUSTER_TYPE" == "rosa" ]] || [[ "$CLUSTER_TYPE" == "hypershift-guest" ]]; then
+        echo "Skip the step. The managed clusters generate the testing accounts by themselves"
+        exit 0
+    fi
+fi
+
+# Main script execution with error handling
+set_proxy || handle_error "set_proxy"
+set_kubeconfig || handle_error "set_kubeconfig"
+check_idp || handle_error "check_idp"
+set_users || handle_error "set_users"
+configure_freeipa || handle_error "configure_freeipa"
+configure_identity_provider || handle_error "configure_identity_provider"
+
+# Installed FreeIPA Version
+echo "FreeIPA Version"
+FREEIPA_POD=$(oc get pods -n freeipa -o jsonpath="{.items[0].metadata.name}" 2>/dev/null)
+if [ -z "$FREEIPA_POD" ]; then
+    echo "No FreeIPA pod found in the 'freeipa' namespace."
+else
+    FREEIPA_VERSION=$(oc exec "$FREEIPA_POD" -n freeipa -- ipa --version)
+    echo "$FREEIPA_VERSION"
+fi
```

Comment:
Maybe it is better to directly print version after previous line `echo "FreeIPA pod is running and ready."`?